pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,426 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servicecontroller
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/golang/glog"
+)
+
+const (
+	clientRetryCount    = 5
+	clientRetryInterval = 5 * time.Second
+)
+
+type ServiceController struct {
+	cloud       cloudprovider.Interface
+	kubeClient  client.Interface
+	clusterName string
+	balancer    cloudprovider.TCPLoadBalancer
+	zone        cloudprovider.Zone
+	mu          sync.Mutex              // protects serviceMap
+	serviceMap  map[string]*api.Service // keys generated by cache.MetaNamespaceKeyFunc
+}
+
+// NewServiceController returns a new service controller to keep cloud provider
+// service resources (like external load balancers) in sync with the registry.
+func NewServiceController(
+	cloud cloudprovider.Interface,
+	kubeClient client.Interface,
+	clusterName string) *ServiceController {
+	return &ServiceController{
+		cloud:       cloud,
+		kubeClient:  kubeClient,
+		clusterName: clusterName,
+		serviceMap:  make(map[string]*api.Service),
+	}
+}
+
+// Run starts a background goroutine that watches for changes to services that
+// have (or had) externalLoadBalancers=true and ensures that they have external
+// load balancers created and deleted appropriately.
+func (s *ServiceController) Run() {
+	if err := s.init(); err != nil {
+		glog.Errorf("Failed to initialized ServiceController: %v", err)
+		return
```

Comment:
This swallows the error, making it indistinguishable from success to the caller.  Does anything other than a log message happen when this error occurs?